diff --git a/dh/__init__.py b/dh/__init__.py
index b91d9c7..3f87013 100644
--- a/dh/__init__.py
+++ b/dh/__init__.py
@@ -17,38 +17,47 @@ C2007CB8 A163BF05 98DA4836 1C55D39A 69163FA8 FD24CF5F
 670C354E 4ABC9804 F1746C08 CA237327 FFFFFFFF FFFFFFFF"""
 # Convert from the value supplied in the RFC to an integer
 prime = read_hex(raw_prime)
-#prime = 13
+
 # Project TODO: write the appropriate code to perform DH key exchange
 
+# A new function to calculate the g_a, 
+# which is the g to the power of a
 def calculate_g_a(g,a):
     g_a = pow(g,a,prime)
     return (g_a)
 
+# The function to create the DH key
 def create_dh_key():
     # Creates a Diffie-Hellman key
     # Returns (public, private)
-	#print("prime:",prime)
+
+    # According to RC3526, the g should be 2
 	g = 2
 	#print("g:",g)
+    
+    # a should be a random number select from the "prime"
 	a = random.randint(1,prime-1)
 	#print("a:",a)
+    
 	# g_a is the g to the power of a mod p
 	g_a = pow(g,a,prime)
 	#print("g-a:",g_a)
 	return (g_a, a)
 
+# The function to calculate DH secret
 def calculate_dh_secret(their_public, my_private):
     # Calculate the shared secret
 	shared_secret = pow(their_public,my_private,prime)
 	#print("shared-secret:",shared_secret)
+    
     # Hash the value so that:
     # (a) There's no bias in the bits of the output
     #     (there may be bias if the shared secret is used raw)
     # (b) We can convert to raw bytes easily
     # (c) We could add additional information if we wanted
     # Feel free to change SHA256 to a different value if more appropriate
-	#test = bytes(str(shared_secret), "ascii")
-	#print(test)
+
+    
 	shared_hash = SHA256.new(bytes(str(shared_secret), "ascii")).hexdigest()
 	#print("shared-hash:",shared_hash)
 	return shared_hash
diff --git a/dh/__pycache__/__init__.cpython-34.pyc b/dh/__pycache__/__init__.cpython-34.pyc
index fe1d12e..939f5ea 100644
Binary files a/dh/__pycache__/__init__.cpython-34.pyc and b/dh/__pycache__/__init__.cpython-34.pyc differ
diff --git a/lib/__pycache__/__init__.cpython-34.pyc b/lib/__pycache__/__init__.cpython-34.pyc
index 78a53b4..4b96cd2 100644
Binary files a/lib/__pycache__/__init__.cpython-34.pyc and b/lib/__pycache__/__init__.cpython-34.pyc differ
diff --git a/lib/__pycache__/comms.cpython-34.pyc b/lib/__pycache__/comms.cpython-34.pyc
index 4764ac5..5571405 100644
Binary files a/lib/__pycache__/comms.cpython-34.pyc and b/lib/__pycache__/comms.cpython-34.pyc differ
diff --git a/lib/__pycache__/evil.cpython-34.pyc b/lib/__pycache__/evil.cpython-34.pyc
index 8cf987f..3998509 100644
Binary files a/lib/__pycache__/evil.cpython-34.pyc and b/lib/__pycache__/evil.cpython-34.pyc differ
diff --git a/lib/__pycache__/files.cpython-34.pyc b/lib/__pycache__/files.cpython-34.pyc
index 1b7591f..fb6050f 100644
Binary files a/lib/__pycache__/files.cpython-34.pyc and b/lib/__pycache__/files.cpython-34.pyc differ
diff --git a/lib/__pycache__/helpers.cpython-34.pyc b/lib/__pycache__/helpers.cpython-34.pyc
index ce5c79f..423925e 100644
Binary files a/lib/__pycache__/helpers.cpython-34.pyc and b/lib/__pycache__/helpers.cpython-34.pyc differ
diff --git a/lib/__pycache__/p2p.cpython-34.pyc b/lib/__pycache__/p2p.cpython-34.pyc
index ae64721..aa0f8b3 100644
Binary files a/lib/__pycache__/p2p.cpython-34.pyc and b/lib/__pycache__/p2p.cpython-34.pyc differ
diff --git a/lib/comms.py b/lib/comms.py
index ceee497..fbe5a5b 100644
--- a/lib/comms.py
+++ b/lib/comms.py
@@ -1,19 +1,18 @@
 import struct
-
-#replay, jiang
-import time
-
 from Crypto.Cipher import XOR
 
-#write by jiang
+# Import to do key exchange, confidentiality, 
+# Integrity and preventing relay attach
 from Crypto.Cipher import AES
 from Crypto import Random
 from Crypto.Hash import HMAC
 from Crypto.Hash import SHA256
-
 from dh import create_dh_key, calculate_dh_secret, calculate_g_a
 
+
 class StealthConn(object):
+    # Initialization. These variable "client", "server" and so on 
+    # will be changed according to the nature of the socket
     def __init__(self, conn, client=False, server=False, verbose=False):
         self.conn = conn
         self.cipher = None
@@ -21,95 +20,111 @@ class StealthConn(object):
         self.server = server
         self.verbose = verbose
 		
-		#give an empty iv, changed by jiang
+		# Use a random number as the initial vector of AES
         self.inital_vector = Random.new().read(AES.block_size)
-		#give an empty aes key
+        
+		# Give a key, which will be asigned value later
         self.aes_key = 0
 		
-        #store the random number
+        # To prevent replay attack, we attach a random number to the 
+        # message, thus the uniqueness of each message could be 
+        # ensured
         self.replay_attack_random = Random.get_random_bytes(14)
-		
-		#prepare a list to store time, jiang
-        #self.time_list = []
-		
-		#prepare a list to store the random
+				
+		# prepare a list to store the random
         self.random_list = []		
 		
+        # initiate_session begins
         self.initiate_session()
 
+        
+        
+    # The initiate session function
     def initiate_session(self):
         # Perform the initial connection handshake for agreeing on a shared secret
-
         ### TODO: Your code here!
         # This can be broken into code run just on the server or just on the client
         if self.server or self.client:
-            #add g,jiang
+        
+            # Create the DH key and store into these two variables
+            # "my_private_key" is a
+            # "my_public_key" is the g_a, 
+            # which means g to the power of a mod p
             my_public_key, my_private_key = create_dh_key()
-			
-            # Send them our public key
+            
+            # Send out public key
+            # For client, it sends g_a as "my_public_key"
             self.send(bytes(str(my_public_key), "ascii"))
-            # Receive their public key
+            
+            # For server, receive public key g_a as "their_public_key"
             their_public_key = int(self.recv())
 			
-            # #send client's g,jiang
-            # if self.client:
-                # self.send(bytes(str(my_public_key), "ascii"))
-                # self.send(bytes(str(g), "ascii"))
-                # their_public_key = int(self.recv())
+            # The following codes was designed to send g. However, according to 
+            # rfc3521, the g should be 2, instead of a random number
+            # Thus the codes has been abandoned
+            '''
+            # send client's g
+            if self.client:
+                self.send(bytes(str(my_public_key), "ascii"))
+                self.send(bytes(str(g), "ascii"))
+                their_public_key = int(self.recv())
 			
-			# #in the server,re calculate g_a,jiang
-            # if self.server:
-                # their_public_key = int(self.recv())
-                # g = int(self.recv())
-                # #print("g:",g)
-                # my_public_key = calculate_g_a(g,my_private_key)
-                # self.send(bytes(str(my_public_key), "ascii"))
-			
-            # Obtain our shared secret
+			in the server,re calculate g_a
+            if self.server:
+                their_public_key = int(self.recv())
+                g = int(self.recv())
+                #print("g:",g)
+                my_public_key = calculate_g_a(g,my_private_key)
+                self.send(bytes(str(my_public_key), "ascii"))
+			'''
+            
+            # Obtain hased shared secret
             shared_hash = calculate_dh_secret(their_public_key, my_private_key)
-            print("Shared hash: {}".format(shared_hash))
+            print("Shared hash: {}".format(shared_hash))        
 
-        # Default XOR algorithm can only take a key of length 32
-        #self.cipher = XOR.new(shared_hash[:4])
-		#the original file use one time padding
-		#jiang change it to aes128, key size 16bytes, mode cfb
+		# The original file use XOR one time padding 
+		# and we changed it to aes128, key size 16bytes, mode cfb       
+        # Becuase the key size is 16 bytes, we take 16 bytes of 
+        # the hashed shared key
         self.aes_key = shared_hash[:16]
-        self.cipher =  AES.new(self.aes_key, AES.MODE_CFB, self.inital_vector)
-
+        
+        # set self.cipher object so that it will enter the "if" in "send" function
+        self.cipher = AES.new(self.aes_key, AES.MODE_CFB, self.inital_vector)
+    
+    # The send function just in client
     def send(self, data):
+        # Only after the encryption is done, the client will send encrypted msg
+        # Or, it will just send the original, as used to send g_a
         if self.cipher:
-            #encrypted_data = self.cipher.encrypt(data)
-			
-			#calculate the encrypted data, jiang change it to aes128, iv is included 
-			#re generate iv, and re generate the cipher
+			# Calculate the encrypted data
+			# Generate initial vector, and the cipher objest
             self.inital_vector = Random.new().read(AES.block_size)
-            self.cipher =  AES.new(self.aes_key, AES.MODE_CFB, self.inital_vector)
+            self.cipher = AES.new(self.aes_key, AES.MODE_CFB, self.inital_vector)
 			
-            #add the time to the start of the message,jiang
-            #if self.client:
-            #ticks = time.time()
-            #ticks = round(ticks, 3)
-			
-			#add 14 random bytes to the start of the message
+			# To prevent replay attack, we add 14 random bytes 
+            # to the start of the message to ensure the uniqueness
             self.replay_attack_random = Random.get_random_bytes(14)
-			
-            #ticks = 1492931133.389
-            #self.replay_attack_random = b'\x9c\xd2\xd2!~\xb9\xb2|\xc3\xc0\x1f(m\xba'
-			
-            #time_string = str(ticks)
-            #with round 3(1ms), bytes_time_string is typically 14 bytes
-            #bytes_time_string = bytes(time_string, "ascii")
             data = self.replay_attack_random + data
 			
+            # Attach the initial vector to the encreypted msg
             encrypted_msg = self.inital_vector + self.cipher.encrypt(data)
+            
+            # Do HMAC to the msg to be sent
             encrypted_msg_hmac = HMAC.new(bytes(self.aes_key,'ascii'), encrypted_msg, SHA256)
             encrypted_msg_hash = encrypted_msg_hmac.hexdigest()
+            
             #print("encrypted_msg_hash:",encrypted_msg_hash)
+            
+            # Attach the hash to the start of encrypted msg to bytes
             encrypted_data = bytes(encrypted_msg_hash,'ascii') + encrypted_msg
+            
+            # Print out to debug
             if self.verbose:
                 print("Original data: {}".format(data))
                 print("Encrypted data: {}".format(repr(encrypted_data)))
                 print("Sending packet of length {}".format(len(encrypted_data)))
+                
+        # If self.cipher is not set, then it means the "send" is just to send g_a
         else:
             encrypted_data = data
 
@@ -117,53 +132,66 @@ class StealthConn(object):
         pkt_len = struct.pack('H', len(encrypted_data))
         self.conn.sendall(pkt_len)
         self.conn.sendall(encrypted_data)
-
+    
+    
+    
+    # The recv function just in server
     def recv(self):
         # Decode the data's length from an unsigned two byte int ('H')
         pkt_len_packed = self.conn.recv(struct.calcsize('H'))
         unpacked_contents = struct.unpack('H', pkt_len_packed)
         pkt_len = unpacked_contents[0]
-
+        
+        # The received data stored in "encrypted_data"
         encrypted_data = self.conn.recv(pkt_len)
+        
+        # Again, only after the encryption is done, the server will do decryption
+        # Otherwise, it will just receive the original, as used to receive g_a
         if self.cipher:
-            #data = self.cipher.decrypt(encrypted_data)
-			
-			#jiang change it to aes decrypt
+        
+			# Firstly, take out the hash to check integrity
             hash = encrypted_data[:64]
-            #print("hash:",hash)
+            #print("hash:",hash)           
             encrypted_msg_hmac = HMAC.new(bytes(self.aes_key,'ascii'), encrypted_data[64:], SHA256)
             encrypted_msg_hash = encrypted_msg_hmac.hexdigest()
+            # encrypted_msg_hash = "abc"
+            
+            # If the hash matches, then do the decryption
+            # Otherwise shut down the connection
             if bytes(encrypted_msg_hash,'ascii') == hash:
-                print("this is from the right sender, because the hash is right")
+                print("Hash is right, message is authentic")
+                
+                # Take out the seperate the initial vector and encrypted msg
                 self.inital_vector = encrypted_data[64:80]
                 encrypted_message = encrypted_data[80:]
+                
+                # Set the self.cipher object
                 self.cipher =  AES.new(self.aes_key, AES.MODE_CFB, self.inital_vector)
-				
-				#handle the random
+                
+                # At this point, the msg consists of the data to send, 
+                # and the random number to prevent replay attack
+                # We need to seperate them
                 random_and_data = self.cipher.decrypt(encrypted_message)
                 send_random_str = random_and_data[:14]
                 data = random_and_data[14:]
+                
+                # Check if it is replayed. If it is, 
+                # then shut down the connection
                 if send_random_str in self.random_list:
                     data = bytes("replay attack","ascii")
                     print("this is replay attack")
+                    
+                # If is not replayed, then append it to the list
                 else:
                     self.random_list.append(send_random_str)
 				
-				#handle the time
-                #time_and_data = self.cipher.decrypt(encrypted_message)
-                #send_time_str = time_and_data[:14]
-                #data = time_and_data[14:]
-                #if send_time_str in self.time_list:
-                    #data = bytes("replay attack","ascii")
-                    #data = "replay attack"
-                    #data = encrypted_data
-                    #print("this is replay attack")
-                #else:
-                    #self.time_list.append(send_time_str)
                 
+            # If hash dose not matches, then shut down the connection
             else:
-                print("this is from the wrong sender, because the hash is not right")
-            
+                print("Hash is wrong, message has been tampered")
+                print("Connection shut down")
+                data = "Hash is wrong, message has been tampered"
+                self.conn.close()
 			
             if self.verbose:
                 print("Receiving packet of length {}".format(pkt_len))
diff --git a/ver 0.11 change log.txt b/ver 0.11 change log.txt
deleted file mode 100644
index b45f1f5..0000000
--- a/ver 0.11 change log.txt	
+++ /dev/null
@@ -1,7 +0,0 @@
-ver 0.11
-添加：
-1. 将g改为永远是2，jiang
-
-ver 0.10
-添加：
-1. 将防replay改为使用random，jiang
\ No newline at end of file
